File: server/src/models/officeConfig.ts
import { Model, DataTypes, Optional } from "sequelize";
import { sequelize } from "@config/database";

export type Point = { latitude: number, longitude: number };

export type GeofenceConfig = {
	included: Point[][];
	excluded: Point[][];
};

interface OfficeConfigAttributes {
	id: number;
	name: string;
	latitude: number;
	longitude: number;
	radius: number; // in meters
	wifiSsid?: string | null;
	geofence?: GeofenceConfig | null;
}

interface OfficeConfigCreationAttributes extends Optional<
	OfficeConfigAttributes,
	"id" | "wifiSsid" | "geofence"
> { }

export default class OfficeConfig
	extends Model<OfficeConfigAttributes, OfficeConfigCreationAttributes>
	implements OfficeConfigAttributes {
	declare public id: number;
	declare public name: string;
	declare public latitude: number;
	declare public longitude: number;
	declare public radius: number;
	declare public wifiSsid: string | null;
	declare public geofence: GeofenceConfig | null;

	declare public readonly createdAt: Date;
	declare public readonly updatedAt: Date;
}

OfficeConfig.init(
	{
		id: {
			type: DataTypes.INTEGER.UNSIGNED,
			autoIncrement: true,
			primaryKey: true,
		},
		name: {
			type: DataTypes.STRING,
			allowNull: false,
		},
		latitude: {
			type: DataTypes.FLOAT,
			allowNull: false,
		},
		longitude: {
			type: DataTypes.FLOAT,
			allowNull: false,
		},
		radius: {
			type: DataTypes.FLOAT,
			allowNull: false,
		},
		wifiSsid: {
			type: DataTypes.STRING,
			allowNull: true,
		},
		geofence: {
			type: DataTypes.JSON,
			allowNull: true,
		},
	},
	{
		sequelize,
		tableName: "office_configs",
		underscored: true,
		timestamps: true,
	},
);

File: client/lib/src/models/office_config.dart
import 'dart:convert';

class GeoPoint {
  final double lat;
  final double lon;

  GeoPoint({required this.lat, required this.lon});

  factory GeoPoint.fromJson(Map<String, dynamic> json) {
    return GeoPoint(
      lat: (json['latitude'] as num).toDouble(),
      lon: (json['longitude'] as num).toDouble(),
    );
  }

  Map<String, dynamic> toJson() => {'latitude': lat, 'longitude': lon};
}

class GeofenceConfig {
  final List<List<GeoPoint>> included;
  final List<List<GeoPoint>> excluded;

  GeofenceConfig({this.included = const [], this.excluded = const []});

  factory GeofenceConfig.fromJson(Map<String, dynamic> json) {
    List<List<GeoPoint>> parsePolygons(dynamic list) {
      if (list == null) return [];
      return (list as List).map((poly) {
        return (poly as List).map((point) => GeoPoint.fromJson(point)).toList();
      }).toList();
    }

    return GeofenceConfig(
      included: parsePolygons(json['included']),
      excluded: parsePolygons(json['excluded']),
    );
  }

  Map<String, dynamic> toJson() {
    return {'included': included, 'excluded': excluded};
  }
}

class OfficeConfig {
  final int? id;
  final String name;
  final double latitude;
  final double longitude;
  final double radius;
  final String? wifiSsid;
  final GeofenceConfig? geofence;

  OfficeConfig({
    this.id,
    required this.name,
    required this.latitude,
    required this.longitude,
    required this.radius,
    this.wifiSsid,
    this.geofence,
  });

  factory OfficeConfig.fromJson(Map<String, dynamic> json) {
    return OfficeConfig(
      id: json['id'],
      name: json['name'] ?? '',
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      radius: (json['radius'] as num).toDouble(),
      wifiSsid: json['wifiSsid'],
      geofence: json['geofence'] != null
          ? GeofenceConfig.fromJson(jsonDecode(json['geofence']))
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'latitude': latitude,
      'longitude': longitude,
      'radius': radius,
      'wifiSsid': wifiSsid,
      'geofence': geofence?.toJson(),
    };
  }
}
